{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Master Table CCLE N28\n",
    "\n",
    "Este notebook contiene los procedimientos realizados para lograr obtener una master table con todas las tablas de reditools correspondientes a las 28 lineas celulares de CCLE.\n",
    "\n",
    "A continuación se muestra el contenido de la carpeta `../data/`, donde se encuentrar los archivos de reditools (`*.tsv`) junto a sus archivos log de reditools (`*_log.txt`):\n",
    "\n",
    "```bash\n",
    "tree ../data/\n",
    ".\n",
    "├── C836.HCC1143.2_log.txt\n",
    "├── C836.HCC1143.2.tsv\n",
    "├── C836.HCC1395.2_log.txt\n",
    "├── C836.HCC1395.2.tsv\n",
    "├── C836.JIMT-1.1_log.txt\n",
    "├── C836.JIMT-1.1.tsv\n",
    "├── C836.MDA-MB-415.1_log.txt\n",
    "├── C836.MDA-MB-415.1.tsv\n",
    "├── G20496.HCC1954.2_log.txt\n",
    "├── G20496.HCC1954.2.tsv\n",
    "├── G26176.CAL-51.2_log.txt\n",
    "├── G26176.CAL-51.2.tsv\n",
    "├── G26185.HCC70.2_log.txt\n",
    "├── G26185.HCC70.2.tsv\n",
    "├── G26206.CAL-120.2_log.txt\n",
    "├── G26206.CAL-120.2.tsv\n",
    "├── G26250.BT-474.2_log.txt\n",
    "├── G26250.BT-474.2.tsv\n",
    "├── G26260.HCC1395.2_log.txt\n",
    "├── G26260.HCC1395.2.tsv\n",
    "├── G27249.AU565.1_log.txt\n",
    "├── G27249.AU565.1.tsv\n",
    "├── G27278.ZR-75-30.1_log.txt\n",
    "├── G27278.ZR-75-30.1.tsv\n",
    "├── G27300.BT-483.1_log.txt\n",
    "├── G27300.BT-483.1.tsv\n",
    "├── G27310.HCC1428.1_log.txt\n",
    "├── G27310.HCC1428.1.tsv\n",
    "├── G27322.HCC1500.1_log.txt\n",
    "├── G27322.HCC1500.1.tsv\n",
    "├── G27334.CAMA-1.1_log.txt\n",
    "├── G27334.CAMA-1.1.tsv\n",
    "├── G27355.EFM-192A.1_log.txt\n",
    "├── G27355.EFM-192A.1.tsv\n",
    "├── G27365.CAL-148.1_log.txt\n",
    "├── G27365.CAL-148.1.tsv\n",
    "├── G28036.JIMT-1.1_log.txt\n",
    "├── G28036.JIMT-1.1.tsv\n",
    "├── G28064.MDA-MB-468.1_log.txt\n",
    "├── G28064.MDA-MB-468.1.tsv\n",
    "├── G28069.MDA-MB-415.1_log.txt\n",
    "├── G28069.MDA-MB-415.1.tsv\n",
    "├── G28072.MDA-MB-175-VII.1_log.txt\n",
    "├── G28072.MDA-MB-175-VII.1.tsv\n",
    "├── G28869.HCC38.3_log.txt\n",
    "├── G28869.HCC38.3.tsv\n",
    "├── G28898.HCC202.3_log.txt\n",
    "├── G28898.HCC202.3.tsv\n",
    "├── G30590.BT-549.1_log.txt\n",
    "├── G30590.BT-549.1.tsv\n",
    "├── G30624.UACC-812.1_log.txt\n",
    "├── G30624.UACC-812.1.tsv\n",
    "├── G41720.HCC2218.5_log.txt\n",
    "├── G41720.HCC2218.5.tsv\n",
    "├── G41743.HCC1419.5_log.txt\n",
    "└── G41743.HCC1419.5.tsv\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PROCEDIMIENTOS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering of samples\n",
    "\n",
    "Filters applied (in order):\n",
    "1. (coverage at position >= 10)\n",
    "2. (var == \"AG\" OR var == \"TC\")\n",
    "3. (coverage of alt. allele >= 5)\n",
    "4. (freq. of alt. allele >= 0.05)               "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering file: G28036.JIMT-1.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26185.HCC70.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: C836.HCC1143.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26176.CAL-51.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28898.HCC202.3.tsv\n",
      "------------------------------------\n",
      "Filtering file: C836.MDA-MB-415.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: C836.HCC1395.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27310.HCC1428.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G30590.BT-549.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: C836.JIMT-1.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G41720.HCC2218.5.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28064.MDA-MB-468.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26260.HCC1395.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28072.MDA-MB-175-VII.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27278.ZR-75-30.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27355.EFM-192A.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27300.BT-483.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26206.CAL-120.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G20496.HCC1954.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27365.CAL-148.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27249.AU565.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G41743.HCC1419.5.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26250.BT-474.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28869.HCC38.3.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27322.HCC1500.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G30624.UACC-812.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28069.MDA-MB-415.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27334.CAMA-1.1.tsv\n",
      "------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Import the libraries\n",
    "import pandas as pd\n",
    "import glob\n",
    "import sys\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "# Function to filter df by coverage and alleles\n",
    "def filtering(path_file, out_dir):\n",
    "    \n",
    "    # Create the out_file name\n",
    "    out_file = os.path.join(out_dir, os.path.basename(path_file).replace('.tsv', '_filtered.tsv'))\n",
    "    \n",
    "    count = 0\n",
    "    with open(path_file, 'r') as fp , open(out_file, 'w+') as fp_out:\n",
    "        for line in fp:\n",
    "            if count == 0:\n",
    "                fp_out.write(f\"{line}\")\n",
    "                count += 1\n",
    "                continue\n",
    "            count += 1\n",
    "            tmp = line.split('\\t')\n",
    "            if (int(tmp[4]) >= 10) & (('AG' in tmp[7]) | ('TC' in tmp[7])):\n",
    "                if ((tmp[2] == 'A') & (int(tmp[6].split(',')[2]) >= 5) & ( int(tmp[6].split(',')[2])/int(tmp[4]) >= 0.05 ) ) | ((tmp[2] == 'T') & (int(tmp[6].split(',')[1]) >= 5) & ( int(tmp[6].split(',')[1])/int(tmp[4]) >= 0.05 )):\n",
    "                    fp_out.write(f\"{line}\")\n",
    "            #if count == 5:\n",
    "            #    break\n",
    "    \n",
    "    print('------------------------------------')\n",
    "\n",
    "\n",
    "# Get list of reditools tables\n",
    "tables = glob.glob(os.path.join('../data', '*.tsv'))\n",
    "\n",
    "# Create the out_dir and out_file\n",
    "out_dir = '../outputs/01_filtering_tables'\n",
    "os.makedirs(out_dir, exist_ok=True)\n",
    "\n",
    "for file in tables:\n",
    "    print('Filtering file: {}'.format(os.path.basename(file)))\n",
    "    filtering(file, out_dir)\n",
    "\n",
    "#filtering(tables[0], out_dir)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hay 4 muestras que se cayeron durante el análisis. Asi que serán eliminadas de la carpeta `01_filtering_tables`. Éstas 4 muestras son las siguientes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Las siguientes muestras deben ser eliminadas:\n",
      "../outputs/01_filtering_tables/C836.JIMT-1.1_filtered.tsv\n",
      "../outputs/01_filtering_tables/C836.MDA-MB-415.1_filtered.tsv\n",
      "../outputs/01_filtering_tables/C836.HCC1395.2_filtered.tsv\n",
      "../outputs/01_filtering_tables/C836.HCC1143.2_filtered.tsv\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Get Samples with errors: \n",
    "echo \"Las siguientes muestras deben ser eliminadas:\"\n",
    "\n",
    "find ../outputs -name \"*.tsv\" | grep -f <(grep -i \"err\" ../data/*_log.txt | cut -f1 -d':' | cut -f3 -d'/' | sed 's/_log.txt//g')\n",
    "\n",
    "# Delete the files\n",
    "rm $(find ../outputs -name \"*.tsv\" | grep -f <(grep -i \"err\" ../data/*_log.txt | cut -f1 -d':' | cut -f3 -d'/' | sed 's/_log.txt//g') | xargs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "El resumen de las variantes filtradas se muestra a continuación:\n",
    "\n",
    "| cell_line               | raw sites | filtered sites |\n",
    "|-------------------------|-----------|----------------|\n",
    "| C836.HCC1143.2          | 223       | NA             |\n",
    "| C836.HCC1395.2          | 505606    | NA             |\n",
    "| C836.JIMT-1.1           | 265409    | NA             |\n",
    "| C836.MDA-MB-415.1       | 608808    | NA             |\n",
    "| G20496.HCC1954.2        | 4180142   | 16442          |\n",
    "| G26176.CAL-51.2         | 4443743   | 24098          |\n",
    "| G26185.HCC70.2          | 4132223   | 20643          |\n",
    "| G26206.CAL-120.2        | 3361752   | 13527          |\n",
    "| G26250.BT-474.2         | 3814955   | 17178          |\n",
    "| G26260.HCC1395.2        | 4112263   | 17521          |\n",
    "| G27249.AU565.1          | 4909995   | 24463          |\n",
    "| G27278.ZR-75-30.1       | 4313095   | 27025          |\n",
    "| G27300.BT-483.1         | 3776925   | 18278          |\n",
    "| G27310.HCC1428.1        | 4095853   | 20480          |\n",
    "| G27322.HCC1500.1        | 4179877   | 29917          |\n",
    "| G27334.CAMA-1.1         | 3850742   | 22348          |\n",
    "| G27355.EFM-192A.1       | 4011857   | 17354          |\n",
    "| G27365.CAL-148.1        | 4554028   | 16963          |\n",
    "| G28036.JIMT-1.1         | 3823738   | 16238          |\n",
    "| G28064.MDA-MB-468.1     | 3499509   | 18532          |\n",
    "| G28069.MDA-MB-415.1     | 4358917   | 26397          |\n",
    "| G28072.MDA-MB-175-VII.1 | 4582056   | 29807          |\n",
    "| G28869.HCC38.3          | 5794550   | 27958          |\n",
    "| G28898.HCC202.3         | 4968291   | 25400          |\n",
    "| G30590.BT-549.1         | 3087657   | 13112          |\n",
    "| G30624.UACC-812.1       | 880975    | 5732           |\n",
    "| G41720.HCC2218.5        | 5187105   | 23774          |\n",
    "| G41743.HCC1419.5        | 5567992   | 25417          |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Multiple Merging\n",
    "The 24 final tables were merged into One Big Master Table (code below)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Strand_G28069.MDA-MB-415.1</th>\n",
       "      <th>Coverage-q30_G28069.MDA-MB-415.1</th>\n",
       "      <th>MeanQ_G28069.MDA-MB-415.1</th>\n",
       "      <th>BaseCount[A,C,G,T]_G28069.MDA-MB-415.1</th>\n",
       "      <th>AllSubs_G28069.MDA-MB-415.1</th>\n",
       "      <th>Frequency_G28069.MDA-MB-415.1</th>\n",
       "      <th>gCoverage-q30_G28069.MDA-MB-415.1</th>\n",
       "      <th>...</th>\n",
       "      <th>Coverage-q30_G20496.HCC1954.2</th>\n",
       "      <th>MeanQ_G20496.HCC1954.2</th>\n",
       "      <th>BaseCount[A,C,G,T]_G20496.HCC1954.2</th>\n",
       "      <th>AllSubs_G20496.HCC1954.2</th>\n",
       "      <th>Frequency_G20496.HCC1954.2</th>\n",
       "      <th>gCoverage-q30_G20496.HCC1954.2</th>\n",
       "      <th>gMeanQ_G20496.HCC1954.2</th>\n",
       "      <th>gBaseCount[A,C,G,T]_G20496.HCC1954.2</th>\n",
       "      <th>gAllSubs_G20496.HCC1954.2</th>\n",
       "      <th>gFrequency_G20496.HCC1954.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14907</td>\n",
       "      <td>A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>36.95</td>\n",
       "      <td>[9, 0, 30, 0]</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.77</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>37.6</td>\n",
       "      <td>[9, 0, 38, 0]</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.81</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15118</td>\n",
       "      <td>A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>35.09</td>\n",
       "      <td>[5, 0, 6, 0]</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.55</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>15447</td>\n",
       "      <td>A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>36.93</td>\n",
       "      <td>[17, 0, 12, 0]</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.41</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16378</td>\n",
       "      <td>T</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>38.15</td>\n",
       "      <td>[0, 12, 0, 1]</td>\n",
       "      <td>TC</td>\n",
       "      <td>0.92</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>38.36</td>\n",
       "      <td>[0, 11, 0, 0]</td>\n",
       "      <td>TC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 267 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Region  Position Reference Strand_G28069.MDA-MB-415.1  \\\n",
       "0      1     14907         A                        2.0   \n",
       "1      1     14930         A                        2.0   \n",
       "2      1     15118         A                        2.0   \n",
       "3      1     15447         A                        2.0   \n",
       "4      1     16378         T                        2.0   \n",
       "\n",
       "  Coverage-q30_G28069.MDA-MB-415.1 MeanQ_G28069.MDA-MB-415.1  \\\n",
       "0                             39.0                     36.95   \n",
       "1                             47.0                      37.6   \n",
       "2                             11.0                     35.09   \n",
       "3                             29.0                     36.93   \n",
       "4                             13.0                     38.15   \n",
       "\n",
       "  BaseCount[A,C,G,T]_G28069.MDA-MB-415.1 AllSubs_G28069.MDA-MB-415.1  \\\n",
       "0                          [9, 0, 30, 0]                          AG   \n",
       "1                          [9, 0, 38, 0]                          AG   \n",
       "2                           [5, 0, 6, 0]                          AG   \n",
       "3                         [17, 0, 12, 0]                          AG   \n",
       "4                          [0, 12, 0, 1]                          TC   \n",
       "\n",
       "  Frequency_G28069.MDA-MB-415.1 gCoverage-q30_G28069.MDA-MB-415.1  ...  \\\n",
       "0                          0.77                                 -  ...   \n",
       "1                          0.81                                 -  ...   \n",
       "2                          0.55                                 -  ...   \n",
       "3                          0.41                                 -  ...   \n",
       "4                          0.92                                 -  ...   \n",
       "\n",
       "  Coverage-q30_G20496.HCC1954.2 MeanQ_G20496.HCC1954.2  \\\n",
       "0                            NA                     NA   \n",
       "1                            NA                     NA   \n",
       "2                            NA                     NA   \n",
       "3                            NA                     NA   \n",
       "4                          11.0                  38.36   \n",
       "\n",
       "  BaseCount[A,C,G,T]_G20496.HCC1954.2 AllSubs_G20496.HCC1954.2  \\\n",
       "0                                  NA                       NA   \n",
       "1                                  NA                       NA   \n",
       "2                                  NA                       NA   \n",
       "3                                  NA                       NA   \n",
       "4                       [0, 11, 0, 0]                       TC   \n",
       "\n",
       "  Frequency_G20496.HCC1954.2 gCoverage-q30_G20496.HCC1954.2  \\\n",
       "0                         NA                             NA   \n",
       "1                         NA                             NA   \n",
       "2                         NA                             NA   \n",
       "3                         NA                             NA   \n",
       "4                        1.0                              -   \n",
       "\n",
       "  gMeanQ_G20496.HCC1954.2 gBaseCount[A,C,G,T]_G20496.HCC1954.2  \\\n",
       "0                      NA                                   NA   \n",
       "1                      NA                                   NA   \n",
       "2                      NA                                   NA   \n",
       "3                      NA                                   NA   \n",
       "4                       -                                    -   \n",
       "\n",
       "  gAllSubs_G20496.HCC1954.2 gFrequency_G20496.HCC1954.2  \n",
       "0                        NA                          NA  \n",
       "1                        NA                          NA  \n",
       "2                        NA                          NA  \n",
       "3                        NA                          NA  \n",
       "4                         -                           -  \n",
       "\n",
       "[5 rows x 267 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Import the libraries \n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import glob\n",
    "from functools import reduce\n",
    "\n",
    "\n",
    "# Table's path\n",
    "tables_path = '../outputs/01_filtering_tables'\n",
    "\n",
    "# Get the table list\n",
    "tables = glob.glob(os.path.join(tables_path, '*.tsv'))\n",
    "# Get a list with columns names\n",
    "cols = 'Region','Position','Reference','Strand','Coverage-q30','MeanQ','BaseCount[A,C,G,T]','AllSubs','Frequency','gCoverage-q30','gMeanQ','gBaseCount[A,C,G,T]','gAllSubs','gFrequency'\n",
    "\n",
    "# Create a list with dataframes\n",
    "tables_df = []\n",
    "for table in tables:\n",
    "    # Open the table\n",
    "    df = pd.read_csv(table, sep=\"\\t\", header=0)\n",
    "    # add_sample_names as column-suffix\n",
    "    id = os.path.basename(table).split('_')[0]\n",
    "    cols_tmp = [col+'_'+id if col not in ['Region', 'Position', 'Reference'] else col for col in cols ]\n",
    "    df.columns = cols_tmp\n",
    "    # Sort the df\n",
    "    df.sort_values(by = ['Region', 'Position'], ascending = [True, True], na_position = 'last', inplace=True)\n",
    "    # Reset index\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    # Change some column types\n",
    "    df = df.astype({'Region':'string','Position':'int64', 'Reference':'string'})\n",
    "    # Append df to list\n",
    "    tables_df.append(df)\n",
    "\n",
    "# Merge the tables\n",
    "tables_merged = reduce(lambda  left,right: pd.merge(left,right,on=['Region', 'Position', 'Reference'], how='outer'), tables_df).fillna('NA')\n",
    " \n",
    "tables_merged.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create outdir\n",
    "os.makedirs('../outputs/02_master_table/', exist_ok=True)\n",
    "# Save the table to Excel\n",
    "tables_merged.to_excel('../outputs/02_master_table/CCLE24_master_table.xlsx', header=True)\n",
    "# Save table to tsv\n",
    "tables_merged.to_csv('../outputs/02_master_table/CCLE24_master_table.tsv', sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Strand_G28069.MDA-MB-415.1</th>\n",
       "      <th>Coverage-q30_G28069.MDA-MB-415.1</th>\n",
       "      <th>BaseCount[A,C,G,T]_G28069.MDA-MB-415.1</th>\n",
       "      <th>Strand_G30624.UACC-812.1</th>\n",
       "      <th>Coverage-q30_G30624.UACC-812.1</th>\n",
       "      <th>BaseCount[A,C,G,T]_G30624.UACC-812.1</th>\n",
       "      <th>Strand_G26260.HCC1395.2</th>\n",
       "      <th>...</th>\n",
       "      <th>BaseCount[A,C,G,T]_G28869.HCC38.3</th>\n",
       "      <th>Strand_G30590.BT-549.1</th>\n",
       "      <th>Coverage-q30_G30590.BT-549.1</th>\n",
       "      <th>BaseCount[A,C,G,T]_G30590.BT-549.1</th>\n",
       "      <th>Strand_G28036.JIMT-1.1</th>\n",
       "      <th>Coverage-q30_G28036.JIMT-1.1</th>\n",
       "      <th>BaseCount[A,C,G,T]_G28036.JIMT-1.1</th>\n",
       "      <th>Strand_G20496.HCC1954.2</th>\n",
       "      <th>Coverage-q30_G20496.HCC1954.2</th>\n",
       "      <th>BaseCount[A,C,G,T]_G20496.HCC1954.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14907</td>\n",
       "      <td>A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>[9, 0, 30, 0]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>[7, 0, 6, 0]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>[9, 0, 38, 0]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>[20, 0, 9, 0]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>[5, 0, 6, 0]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15118</td>\n",
       "      <td>A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>[5, 0, 6, 0]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>15447</td>\n",
       "      <td>A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>[17, 0, 12, 0]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16378</td>\n",
       "      <td>T</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>[0, 12, 0, 1]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>[0, 7, 0, 10]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 11, 0, 1]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>[0, 12, 0, 7]</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>[0, 11, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136242</th>\n",
       "      <td>X</td>\n",
       "      <td>133561561</td>\n",
       "      <td>T</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>[0, 6, 0, 96]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136243</th>\n",
       "      <td>X</td>\n",
       "      <td>135046009</td>\n",
       "      <td>A</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>[1, 0, 132, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136244</th>\n",
       "      <td>X</td>\n",
       "      <td>148678982</td>\n",
       "      <td>A</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>[0, 0, 23, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136245</th>\n",
       "      <td>X</td>\n",
       "      <td>149997252</td>\n",
       "      <td>A</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>[0, 0, 13, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136246</th>\n",
       "      <td>X</td>\n",
       "      <td>152958947</td>\n",
       "      <td>T</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>[0, 112, 0, 1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136247 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Region   Position Reference Strand_G28069.MDA-MB-415.1  \\\n",
       "0           1      14907         A                        2.0   \n",
       "1           1      14930         A                        2.0   \n",
       "2           1      15118         A                        2.0   \n",
       "3           1      15447         A                        2.0   \n",
       "4           1      16378         T                        2.0   \n",
       "...       ...        ...       ...                        ...   \n",
       "136242      X  133561561         T                         NA   \n",
       "136243      X  135046009         A                         NA   \n",
       "136244      X  148678982         A                         NA   \n",
       "136245      X  149997252         A                         NA   \n",
       "136246      X  152958947         T                         NA   \n",
       "\n",
       "       Coverage-q30_G28069.MDA-MB-415.1  \\\n",
       "0                                  39.0   \n",
       "1                                  47.0   \n",
       "2                                  11.0   \n",
       "3                                  29.0   \n",
       "4                                  13.0   \n",
       "...                                 ...   \n",
       "136242                               NA   \n",
       "136243                               NA   \n",
       "136244                               NA   \n",
       "136245                               NA   \n",
       "136246                               NA   \n",
       "\n",
       "       BaseCount[A,C,G,T]_G28069.MDA-MB-415.1 Strand_G30624.UACC-812.1  \\\n",
       "0                               [9, 0, 30, 0]                       NA   \n",
       "1                               [9, 0, 38, 0]                       NA   \n",
       "2                                [5, 0, 6, 0]                       NA   \n",
       "3                              [17, 0, 12, 0]                       NA   \n",
       "4                               [0, 12, 0, 1]                      2.0   \n",
       "...                                       ...                      ...   \n",
       "136242                                     NA                       NA   \n",
       "136243                                     NA                       NA   \n",
       "136244                                     NA                       NA   \n",
       "136245                                     NA                       NA   \n",
       "136246                                     NA                       NA   \n",
       "\n",
       "       Coverage-q30_G30624.UACC-812.1 BaseCount[A,C,G,T]_G30624.UACC-812.1  \\\n",
       "0                                  NA                                   NA   \n",
       "1                                  NA                                   NA   \n",
       "2                                  NA                                   NA   \n",
       "3                                  NA                                   NA   \n",
       "4                                17.0                        [0, 7, 0, 10]   \n",
       "...                               ...                                  ...   \n",
       "136242                             NA                                   NA   \n",
       "136243                             NA                                   NA   \n",
       "136244                             NA                                   NA   \n",
       "136245                             NA                                   NA   \n",
       "136246                             NA                                   NA   \n",
       "\n",
       "       Strand_G26260.HCC1395.2  ... BaseCount[A,C,G,T]_G28869.HCC38.3  \\\n",
       "0                           NA  ...                                NA   \n",
       "1                           NA  ...                     [20, 0, 9, 0]   \n",
       "2                           NA  ...                                NA   \n",
       "3                           NA  ...                                NA   \n",
       "4                          2.0  ...                     [0, 11, 0, 1]   \n",
       "...                        ...  ...                               ...   \n",
       "136242                      NA  ...                                NA   \n",
       "136243                      NA  ...                                NA   \n",
       "136244                      NA  ...                                NA   \n",
       "136245                      NA  ...                                NA   \n",
       "136246                      NA  ...                                NA   \n",
       "\n",
       "       Strand_G30590.BT-549.1 Coverage-q30_G30590.BT-549.1  \\\n",
       "0                         2.0                         13.0   \n",
       "1                         2.0                         11.0   \n",
       "2                          NA                           NA   \n",
       "3                          NA                           NA   \n",
       "4                         2.0                         19.0   \n",
       "...                       ...                          ...   \n",
       "136242                     NA                           NA   \n",
       "136243                     NA                           NA   \n",
       "136244                     NA                           NA   \n",
       "136245                     NA                           NA   \n",
       "136246                     NA                           NA   \n",
       "\n",
       "       BaseCount[A,C,G,T]_G30590.BT-549.1 Strand_G28036.JIMT-1.1  \\\n",
       "0                            [7, 0, 6, 0]                     NA   \n",
       "1                            [5, 0, 6, 0]                     NA   \n",
       "2                                      NA                     NA   \n",
       "3                                      NA                     NA   \n",
       "4                           [0, 12, 0, 7]                     NA   \n",
       "...                                   ...                    ...   \n",
       "136242                                 NA                     NA   \n",
       "136243                                 NA                     NA   \n",
       "136244                                 NA                     NA   \n",
       "136245                                 NA                     NA   \n",
       "136246                                 NA                     NA   \n",
       "\n",
       "       Coverage-q30_G28036.JIMT-1.1 BaseCount[A,C,G,T]_G28036.JIMT-1.1  \\\n",
       "0                                NA                                 NA   \n",
       "1                                NA                                 NA   \n",
       "2                                NA                                 NA   \n",
       "3                                NA                                 NA   \n",
       "4                                NA                                 NA   \n",
       "...                             ...                                ...   \n",
       "136242                           NA                                 NA   \n",
       "136243                           NA                                 NA   \n",
       "136244                           NA                                 NA   \n",
       "136245                           NA                                 NA   \n",
       "136246                           NA                                 NA   \n",
       "\n",
       "       Strand_G20496.HCC1954.2 Coverage-q30_G20496.HCC1954.2  \\\n",
       "0                           NA                            NA   \n",
       "1                           NA                            NA   \n",
       "2                           NA                            NA   \n",
       "3                           NA                            NA   \n",
       "4                          2.0                          11.0   \n",
       "...                        ...                           ...   \n",
       "136242                     2.0                         102.0   \n",
       "136243                     2.0                         133.0   \n",
       "136244                     2.0                          23.0   \n",
       "136245                     2.0                          13.0   \n",
       "136246                     2.0                         113.0   \n",
       "\n",
       "       BaseCount[A,C,G,T]_G20496.HCC1954.2  \n",
       "0                                       NA  \n",
       "1                                       NA  \n",
       "2                                       NA  \n",
       "3                                       NA  \n",
       "4                            [0, 11, 0, 0]  \n",
       "...                                    ...  \n",
       "136242                       [0, 6, 0, 96]  \n",
       "136243                      [1, 0, 132, 0]  \n",
       "136244                       [0, 0, 23, 0]  \n",
       "136245                       [0, 0, 13, 0]  \n",
       "136246                      [0, 112, 0, 1]  \n",
       "\n",
       "[136247 rows x 75 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create minimalist table\n",
    "# Get a list with columns names\n",
    "cols = tables_merged.columns\n",
    "# Keep only the minimal columns\n",
    "cols2keep = ['Region', 'Position', 'Reference']+[col for col in cols if (('Coverage-q30' in col) | ('BaseCount[A,C,G,T]' in col) | ('Strand' in col)) ]\n",
    "# Remove the \"g\" columns\n",
    "cols2keep = [col for col in cols2keep if (('gBaseCount[A,C,G,T]' not in col) & ('gCoverage-q30' not in col)) ]\n",
    "\n",
    "# Create the minimal table and save it to file\n",
    "tables_merged_minimal = tables_merged[cols2keep].copy()\n",
    "\n",
    "# Save to file\n",
    "tables_merged_minimal.to_excel('../outputs/02_master_table/CCLE24_master_table_minimal.xlsx', header=True)\n",
    "# Save table to tsv\n",
    "tables_merged_minimal.to_csv('../outputs/02_master_table/CCLE24_master_table_minimal.tsv', sep='\\t', header=True, index=False)\n",
    "\n",
    "# Preview table\n",
    "tables_merged_minimal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Counting reads with bam-readcount\n",
    "\n",
    "GitHub of the project: [bam-readcount](https://github.com/genome/bam-readcount)\n",
    "\n",
    "Tutorial: [bam-readcount Tutorial](https://github.com/genome/bam-readcount/tree/master/tutorial)\n",
    "\n",
    "Installation of the software was donde following github instructions for compilig method on YB's VM.\n",
    "\n",
    "## Create the regions list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# table with editions\n",
    "editions=\"../outputs/02_master_table/CCLE24_master_table_minimal.tsv\"\n",
    "awk 'NR>1{print $1\"\\t\"$2\"\\t\"$2}' $editions > ../outputs/02_master_table/CCLE24_master_table_minimal.regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run bam-readcount\n",
    "`bam-readcount` fue ejecutado para cada linea celular, utilizando la lista de posiciones creada en el paso anterior.\n",
    "\n",
    "El comando utilizado se muestra a continuación:\n",
    "\n",
    "```bash\n",
    "    bam-readcount -b 30 -q 20 -w1 -f ${reference_fasta} -l ${regions_list} ${bam_file} > ${outdir}/${sample}.tsv\n",
    "```\n",
    "\n",
    "Donde:\n",
    "\n",
    "  * -b 30: Minimun base quality of 30.\n",
    "  * -q 20: Minumun mapping quality of 20.\n",
    "  * -w1: to show warnings just once.\n",
    "  * -f: Genoma de referencia. En este caso se utilizó el archivo del repositorio del Broad Institute (Link al archivo)[https://data.broadinstitute.org/snowman/hg19/Homo_sapiens_assembly19.fasta].\n",
    "  * -l: Es la lista de posiciones de interés.\n",
    "\n",
    "Esto se ejecutó mediante los scripts `bam-readcount_parallel.sh` y `bam-readcount.sh`.\n",
    "\n",
    "```bash\n",
    "\n",
    "cd ../code\n",
    "./bam-readcount_parallel.sh\n",
    "\n",
    "```\n",
    "\n",
    "Algunas aclaraciones:\n",
    "\n",
    "- El comando anterior ejecutará el script analizando 8 muestras a la vez. Si se desea cambiar este valor, se debe modificar la variable `MAX_NPROC=8`.\n",
    "- La carpeta de input y output están asignadas directamente en las lineas:  \n",
    "     50: `find ../../CCLE_N28/bams -name \"*.bam\" | while read file;`  \n",
    "     44: `out_dir=\"../outputs/04_bam-readcount\"` \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## bam-readcount Results Post-processing.\n",
    "\n",
    "Se dejará solamente la cobertura total del sitio y de cada alelo, junto a su respectiva cobertura.   \n",
    "De acuerdo a la documentación, las columnas de output representan lo siguiente:\n",
    "\n",
    "> Output\n",
    "> ------\n",
    "> \n",
    "> Output is tab-separated with no header to `STDOUT`, one line per\n",
    "> position:\n",
    "> \n",
    ">     chr\tposition\treference_base\tdepth\tbase:count:avg_mapping_quality:avg_basequality:avg_se_mapping_quality:num_plus_strand:num_minus_strand:avg_pos_as_fraction:avg_num_mismatches_as_fraction:avg_sum_mismatch_qualities:num_q2_containing_reads:avg_distance_to_q2_start_in_q2_reads:avg_clipped_length:avg_distance_to_effective_3p_end   ...\n",
    "> \n",
    "> There is one set of `:`-separated fields for each reported `base` with\n",
    "> statistics on the set of reads containing that base:\n",
    "> \n",
    "> Field | Description\n",
    "> ----- | -----------\n",
    "> base | The base, eg `C`\n",
    "> count | Number of reads\n",
    "> avg_mapping_quality | Mean mapping quality\n",
    "> avg_basequality | Mean base quality\n",
    "> avg_se_mapping_quality | Mean single ended mapping quality\n",
    "> num_plus_strand | Number of reads on the plus/forward strand\n",
    "> num_minus_strand | Number of reads on the minus/reverse strand\n",
    "> avg_pos_as_fraction | Average position on the read as a fraction, calculated with respect to the length after clipping. This value is normalized to the center of the read: bases occurring strictly at the center of the read have a value of 1, those occurring strictly at the ends should approach a value of 0\n",
    "> avg_num_mismatches_as_fraction | Average number of mismatches on these reads per base\n",
    "> avg_sum_mismatch_qualities | Average sum of the base qualities of mismatches in the reads\n",
    "> num_q2_containing_reads | Number of reads with q2 runs at the 3’ end\n",
    "> avg_distance_to_q2_start_in_q2_reads | Average distance of position (as fraction of unclipped read length) to the start of the q2 run\n",
    "> avg_clipped_length | Average clipped read length\n",
    "> avg_distance_to_effective_3p_end | Average distance to the 3’ prime end of the read (as fraction of unclipped read length)\n",
    "\n",
    "\n",
    "\n",
    "Sólo nos quedaremos con:  \n",
    "        `chr position\treference_base\tdepth   base:count`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering file: G28036.JIMT-1.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26185.HCC70.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: C836.HCC1143.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26176.CAL-51.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28898.HCC202.3.tsv\n",
      "------------------------------------\n",
      "Filtering file: C836.MDA-MB-415.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: C836.HCC1395.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27310.HCC1428.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G30590.BT-549.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: C836.JIMT-1.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G41720.HCC2218.5.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28064.MDA-MB-468.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26260.HCC1395.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28072.MDA-MB-175-VII.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27278.ZR-75-30.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27355.EFM-192A.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27300.BT-483.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26206.CAL-120.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G20496.HCC1954.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27365.CAL-148.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27249.AU565.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G41743.HCC1419.5.tsv\n",
      "------------------------------------\n",
      "Filtering file: G26250.BT-474.2.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28869.HCC38.3.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27322.HCC1500.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G30624.UACC-812.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G28069.MDA-MB-415.1.tsv\n",
      "------------------------------------\n",
      "Filtering file: G27334.CAMA-1.1.tsv\n",
      "------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Import the libraries\n",
    "import pandas as pd\n",
    "import glob\n",
    "import sys\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "# Function to get coverage per site and per nucleotide from bam-readcount output table\n",
    "def reduce_table(path_table, out_dir):\n",
    "    \n",
    "    # Create the out_file name\n",
    "    out_file = os.path.join(out_dir, os.path.basename(path_table).replace('.tsv', '_minimal.tsv'))\n",
    "    \n",
    "    count = 0\n",
    "    with open(path_table, 'r') as fp , open(out_file, 'w+') as fp_out:\n",
    "        for line in fp:\n",
    "            count += 1\n",
    "            tmp_line = line.split('\\t')\n",
    "            tmpA = tmp_line[5].split(':')[1]    # A counts\n",
    "            tmpC = tmp_line[6].split(':')[1]    # C counts\n",
    "            tmpG = tmp_line[7].split(':')[1]    # G counts\n",
    "            tmpT = tmp_line[8].split(':')[1]    # T counts\n",
    "            # Generate out_line with the base count in the same format as reditools ([A,C,G,T])\n",
    "            out_line = '{}\\t{}\\t{}\\t{}\\t[{},{},{},{}]\\n'.format(tmp_line[0], tmp_line[1], tmp_line[2], tmp_line[3], tmpA, tmpC, tmpG, tmpT)\n",
    "            # write to output file\n",
    "            fp_out.write(f\"{out_line}\")\n",
    "            \n",
    "    print('------------------------------------')\n",
    "\n",
    "\n",
    "# Get list of bam-readcount tables\n",
    "tables = glob.glob(os.path.join('../outputs/04_bam-readcount', '*.tsv'))\n",
    "\n",
    "# set out_dir\n",
    "out_dir = '../outputs/04_bam-readcount'\n",
    "\n",
    "for file in tables:\n",
    "    print('Reducing file: {}'.format(os.path.basename(file)))\n",
    "    reduce_table(file, out_dir)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge Master-Table w/ bam-readcount tables\n",
    "El último paso es unir la master table `CCLE24_master_table_minimal.tsv`, con la información obtenida del software `bam-readcount`.\n",
    "\n",
    "La Tabla a continuación muestra el número de ediciones que fueron recuperadas en cada línea celular.\n",
    "\n",
    "Código usado para obtener los conteos:\n",
    "\n",
    "```bash\n",
    "    wc -l ../outputs/04_bam-readcount/*minimal.tsv  | sed 's/..\\/outputs\\/04_bam-readcount\\///g' | awk '{print $2\"\\t\"$1}' \n",
    "\n",
    "```\n",
    "\n",
    "| Cell line | Number of editions |\n",
    "|---|---|\n",
    "| C836.HCC1143.2_minimal.tsv | 36320 |\n",
    "| C836.HCC1395.2_minimal.tsv | 37883 |\n",
    "| C836.JIMT-1.1_minimal.tsv | 36030 |\n",
    "| C836.MDA-MB-415.1_minimal.tsv | 36606 |\n",
    "| G20496.HCC1954.2_minimal.tsv | 124523 |\n",
    "| G26176.CAL-51.2_minimal.tsv | 122034 |\n",
    "| G26185.HCC70.2_minimal.tsv | 121980 |\n",
    "| G26206.CAL-120.2_minimal.tsv | 120415 |\n",
    "| G26250.BT-474.2_minimal.tsv | 124505 |\n",
    "| G26260.HCC1395.2_minimal.tsv | 122344 |\n",
    "| G27249.AU565.1_minimal.tsv | 123222 |\n",
    "| G27278.ZR-75-30.1_minimal.tsv | 123092 |\n",
    "| G27300.BT-483.1_minimal.tsv | 126672 |\n",
    "| G27310.HCC1428.1_minimal.tsv | 126807 |\n",
    "| G27322.HCC1500.1_minimal.tsv | 126778 |\n",
    "| G27334.CAMA-1.1_minimal.tsv | 124416 |\n",
    "| G27355.EFM-192A.1_minimal.tsv | 126402 |\n",
    "| G27365.CAL-148.1_minimal.tsv | 123054 |\n",
    "| G28036.JIMT-1.1_minimal.tsv | 122959 |\n",
    "| G28064.MDA-MB-468.1_minimal.tsv | 124777 |\n",
    "| G28069.MDA-MB-415.1_minimal.tsv | 125989 |\n",
    "| G28072.MDA-MB-175-VII.1_minimal.tsv | 126914 |\n",
    "| G28869.HCC38.3_minimal.tsv | 127695 |\n",
    "| G28898.HCC202.3_minimal.tsv | 124223 |\n",
    "| G30590.BT-549.1_minimal.tsv | 117085 |\n",
    "| G30624.UACC-812.1_minimal.tsv | 114631 |\n",
    "| G41720.HCC2218.5_minimal.tsv | 123190 |\n",
    "| G41743.HCC1419.5_minimal.tsv | 124882 |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_5281/1132225235.py:12: DtypeWarning: Columns (0,5,8,11,14,17,20,23,26,29,32,35,38,41,44,47,50,53,56,59,62) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mt_df = pd.read_csv(mt, sep=\"\\t\", header=0)\n"
     ]
    }
   ],
   "source": [
    "# Import libraries\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import glob\n",
    "import numpy as np\n",
    "from functools import reduce   # Import reduce function\n",
    "\n",
    "# Master table CCLE24\n",
    "mt = '../outputs/02_master_table/CCLE24_master_table_minimal.tsv'\n",
    "# Read the MT\n",
    "mt_df = pd.read_csv(mt, sep=\"\\t\", header=0)\n",
    "\n",
    "# bam-readcount tables\n",
    "tables = glob.glob(os.path.join('../outputs/04_bam-readcount', '*minimal.tsv'))\n",
    "# remove the C836 files (corrupt files)\n",
    "tables = [x for x in tables if 'C836' not in x]\n",
    "\n",
    "# Create a copy of MT\n",
    "mt_df2 = mt_df.copy()\n",
    "# Sort the df\n",
    "mt_df2.sort_values(by = ['Region', 'Position'], ascending = [True, True], na_position = 'last', inplace=True)\n",
    "# Reset index\n",
    "mt_df2.reset_index(drop=True, inplace=True)\n",
    "# Change column types\n",
    "mt_df2 = mt_df2.astype({'Region':'string','Position':'int64', 'Reference':'string'})\n",
    "\n",
    "# Merge each bam-readcount table with the MT\n",
    "for table in tables:\n",
    "    # Open the table. Columns will be named as: Region  Position    Reference   bam-rc_Coverage-q20    bam-rc[A,C,G,T]\n",
    "    df = pd.read_csv(table, sep=\"\\t\", header=None, names=['Region','Position','Reference','bam-rc_Coverage-q20','bam-rc[A,C,G,T]'])\n",
    "    # add_sample_names as column-suffix\n",
    "    id = os.path.basename(table).split('_')[0]\n",
    "    cols_tmp = [ col+'_'+id if col not in ['Region', 'Position', 'Reference'] else col for col in list(df.columns) ]\n",
    "    df.columns = cols_tmp\n",
    "    # Sort the df\n",
    "    df.sort_values(by = ['Region', 'Position'], ascending = [True, True], na_position = 'last', inplace=True)\n",
    "    # Reset index\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    # Change some column types\n",
    "    df = df.astype({'Region':'string','Position':'int64', 'Reference':'string'})\n",
    "    # perform the merge\n",
    "    mt_df2 = mt_df2.merge(df, on=['Region','Position','Reference'], how='left')\n",
    "\n",
    "# Reorder columns\n",
    "cols = list(mt_df2.columns)\n",
    "cols_samples = [col for col in cols if col not in ['Region', 'Position', 'Reference'] ]\n",
    "\n",
    "# Get uniq ids\n",
    "ids = [x.split('_')[-1] for x in cols_samples]\n",
    "ids = list(set(ids))\n",
    "\n",
    "# Reorder sample cols\n",
    "cols_new_order = ['Region', 'Position', 'Reference']\n",
    "for id in ids:\n",
    "    tmp_cols = [x for x in cols_samples if id in x]\n",
    "    cols_new_order = cols_new_order + tmp_cols\n",
    "\n",
    "# final MT\n",
    "mt_df_final = mt_df2[cols_new_order]\n",
    "\n",
    "# Save the final table\n",
    "outdir='../outputs/05_mt-bamreadcount_merge/'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "mt_df_final.to_csv(os.path.join(outdir, 'CCL24_master_table_minimal_with_bamreadcount.tsv'), sep=\"\\t\", header=True, index=False, na_rep='NA')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output a minimal version, just with {Region}, {Position}, {Reference}, {BaseCount[A,C,G,T]_*} and {bam-rc[A,C,G,T]_*}\n",
    "minim_cols = [x for x in mt_df_final.columns if not x.startswith('Strand') and not x.startswith('Coverage') and not x.startswith('bam-rc_Coverage')]\n",
    "# Keep the minimun columns\n",
    "mt_df_final_minimal = mt_df_final[minim_cols]\n",
    "# Save to table\n",
    "mt_df_final_minimal.to_csv(os.path.join(outdir, 'CCL24_master_table_just-counts_with_bamreadcount.tsv'), sep=\"\\t\", header=True, index=False, na_rep='NA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference</th>\n",
       "      <th>BaseCount[A,C,G,T]_G27334.CAMA-1.1</th>\n",
       "      <th>bam-rc[A,C,G,T]_G27334.CAMA-1.1</th>\n",
       "      <th>BaseCount[A,C,G,T]_G26206.CAL-120.2</th>\n",
       "      <th>bam-rc[A,C,G,T]_G26206.CAL-120.2</th>\n",
       "      <th>BaseCount[A,C,G,T]_G27310.HCC1428.1</th>\n",
       "      <th>bam-rc[A,C,G,T]_G27310.HCC1428.1</th>\n",
       "      <th>BaseCount[A,C,G,T]_G27249.AU565.1</th>\n",
       "      <th>...</th>\n",
       "      <th>BaseCount[A,C,G,T]_G26260.HCC1395.2</th>\n",
       "      <th>bam-rc[A,C,G,T]_G26260.HCC1395.2</th>\n",
       "      <th>BaseCount[A,C,G,T]_G28072.MDA-MB-175-VII.1</th>\n",
       "      <th>bam-rc[A,C,G,T]_G28072.MDA-MB-175-VII.1</th>\n",
       "      <th>BaseCount[A,C,G,T]_G30624.UACC-812.1</th>\n",
       "      <th>bam-rc[A,C,G,T]_G30624.UACC-812.1</th>\n",
       "      <th>BaseCount[A,C,G,T]_G20496.HCC1954.2</th>\n",
       "      <th>bam-rc[A,C,G,T]_G20496.HCC1954.2</th>\n",
       "      <th>BaseCount[A,C,G,T]_G28069.MDA-MB-415.1</th>\n",
       "      <th>bam-rc[A,C,G,T]_G28069.MDA-MB-415.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14907</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[6,0,3,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[9,0,3,0]</td>\n",
       "      <td>[4, 0, 12, 0]</td>\n",
       "      <td>[4,0,12,0]</td>\n",
       "      <td>[6, 0, 13, 0]</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[5,0,1,0]</td>\n",
       "      <td>[10, 0, 10, 0]</td>\n",
       "      <td>[10,0,10,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[4,0,2,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,5,0]</td>\n",
       "      <td>[9, 0, 30, 0]</td>\n",
       "      <td>[9,0,30,0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14930</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[8,0,3,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[9,0,2,0]</td>\n",
       "      <td>[4, 0, 15, 0]</td>\n",
       "      <td>[4,0,15,0]</td>\n",
       "      <td>[6, 0, 11, 0]</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[5,0,1,0]</td>\n",
       "      <td>[8, 0, 19, 0]</td>\n",
       "      <td>[8,0,19,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[4,0,1,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1,0,6,0]</td>\n",
       "      <td>[9, 0, 38, 0]</td>\n",
       "      <td>[9,0,38,0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15118</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[1,0,1,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[2,0,0,0]</td>\n",
       "      <td>[5, 0, 6, 0]</td>\n",
       "      <td>[5,0,6,0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>15447</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[6,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[10,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[7,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[9,0,1,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[19,0,2,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[4,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[12,0,0,0]</td>\n",
       "      <td>[17, 0, 12, 0]</td>\n",
       "      <td>[17,0,12,0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16378</td>\n",
       "      <td>T</td>\n",
       "      <td>[0, 15, 0, 13]</td>\n",
       "      <td>[0,15,0,14]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,3,0,12]</td>\n",
       "      <td>[0, 8, 0, 8]</td>\n",
       "      <td>[0,8,0,8]</td>\n",
       "      <td>[0, 28, 0, 15]</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 12, 0, 6]</td>\n",
       "      <td>[0,13,0,6]</td>\n",
       "      <td>[0, 27, 0, 19]</td>\n",
       "      <td>[0,27,0,19]</td>\n",
       "      <td>[0, 7, 0, 10]</td>\n",
       "      <td>[0,7,0,10]</td>\n",
       "      <td>[0, 11, 0, 0]</td>\n",
       "      <td>[0,11,0,0]</td>\n",
       "      <td>[0, 12, 0, 1]</td>\n",
       "      <td>[0,12,0,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136242</th>\n",
       "      <td>Y</td>\n",
       "      <td>21153474</td>\n",
       "      <td>A</td>\n",
       "      <td>[2, 0, 1033, 0]</td>\n",
       "      <td>[2,0,1048,0]</td>\n",
       "      <td>[0, 0, 110, 0]</td>\n",
       "      <td>[0,0,117,0]</td>\n",
       "      <td>[1, 0, 713, 0]</td>\n",
       "      <td>[1,0,724,0]</td>\n",
       "      <td>[3, 3, 2119, 1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 0, 24, 0]</td>\n",
       "      <td>[0,0,30,0]</td>\n",
       "      <td>[1, 1, 1304, 0]</td>\n",
       "      <td>[1,1,1306,0]</td>\n",
       "      <td>[0, 0, 370, 0]</td>\n",
       "      <td>[0,0,373,0]</td>\n",
       "      <td>[1, 0, 2900, 1]</td>\n",
       "      <td>[1,0,2903,1]</td>\n",
       "      <td>[5, 2, 2367, 1]</td>\n",
       "      <td>[5,2,2369,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136243</th>\n",
       "      <td>Y</td>\n",
       "      <td>21153609</td>\n",
       "      <td>T</td>\n",
       "      <td>[2, 2077, 0, 1]</td>\n",
       "      <td>[2,2088,0,1]</td>\n",
       "      <td>[0, 257, 0, 0]</td>\n",
       "      <td>[0,266,0,0]</td>\n",
       "      <td>[0, 1405, 0, 1]</td>\n",
       "      <td>[0,1415,0,1]</td>\n",
       "      <td>[0, 3837, 0, 1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 90, 0, 0]</td>\n",
       "      <td>[0,94,0,0]</td>\n",
       "      <td>[0, 2296, 0, 0]</td>\n",
       "      <td>[0,2297,0,0]</td>\n",
       "      <td>[0, 445, 0, 0]</td>\n",
       "      <td>[0,451,0,0]</td>\n",
       "      <td>[1, 4965, 0, 2]</td>\n",
       "      <td>[1,4970,0,2]</td>\n",
       "      <td>[0, 4015, 0, 0]</td>\n",
       "      <td>[0,4019,0,0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136244</th>\n",
       "      <td>Y</td>\n",
       "      <td>21154569</td>\n",
       "      <td>A</td>\n",
       "      <td>[1, 0, 2462, 0]</td>\n",
       "      <td>[1,0,2465,0]</td>\n",
       "      <td>[0, 0, 453, 0]</td>\n",
       "      <td>[0,0,458,0]</td>\n",
       "      <td>[0, 0, 1848, 0]</td>\n",
       "      <td>[0,0,1850,0]</td>\n",
       "      <td>[3, 1, 6142, 1]</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 0, 165, 0]</td>\n",
       "      <td>[0,0,169,0]</td>\n",
       "      <td>[1, 0, 3501, 1]</td>\n",
       "      <td>[1,0,3502,1]</td>\n",
       "      <td>[0, 0, 565, 0]</td>\n",
       "      <td>[0,0,566,0]</td>\n",
       "      <td>[0, 0, 5800, 4]</td>\n",
       "      <td>[0,0,5802,4]</td>\n",
       "      <td>[3, 0, 4944, 0]</td>\n",
       "      <td>[3,0,4944,0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136245</th>\n",
       "      <td>Y</td>\n",
       "      <td>21154706</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,0,1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,1,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,0,7]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,1,0,4]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,0,13]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136246</th>\n",
       "      <td>Y</td>\n",
       "      <td>26328502</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,1,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0,0,0,0]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136247 rows × 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Region  Position Reference BaseCount[A,C,G,T]_G27334.CAMA-1.1  \\\n",
       "0           1     14907         A                                NaN   \n",
       "1           1     14930         A                                NaN   \n",
       "2           1     15118         A                                NaN   \n",
       "3           1     15447         A                                NaN   \n",
       "4           1     16378         T                     [0, 15, 0, 13]   \n",
       "...       ...       ...       ...                                ...   \n",
       "136242      Y  21153474         A                    [2, 0, 1033, 0]   \n",
       "136243      Y  21153609         T                    [2, 2077, 0, 1]   \n",
       "136244      Y  21154569         A                    [1, 0, 2462, 0]   \n",
       "136245      Y  21154706         T                                NaN   \n",
       "136246      Y  26328502         A                                NaN   \n",
       "\n",
       "       bam-rc[A,C,G,T]_G27334.CAMA-1.1 BaseCount[A,C,G,T]_G26206.CAL-120.2  \\\n",
       "0                            [6,0,3,0]                                 NaN   \n",
       "1                            [8,0,3,0]                                 NaN   \n",
       "2                            [0,0,0,0]                                 NaN   \n",
       "3                            [6,0,0,0]                                 NaN   \n",
       "4                          [0,15,0,14]                                 NaN   \n",
       "...                                ...                                 ...   \n",
       "136242                    [2,0,1048,0]                      [0, 0, 110, 0]   \n",
       "136243                    [2,2088,0,1]                      [0, 257, 0, 0]   \n",
       "136244                    [1,0,2465,0]                      [0, 0, 453, 0]   \n",
       "136245                       [0,0,0,1]                                 NaN   \n",
       "136246                             NaN                                 NaN   \n",
       "\n",
       "       bam-rc[A,C,G,T]_G26206.CAL-120.2 BaseCount[A,C,G,T]_G27310.HCC1428.1  \\\n",
       "0                             [9,0,3,0]                       [4, 0, 12, 0]   \n",
       "1                             [9,0,2,0]                       [4, 0, 15, 0]   \n",
       "2                             [1,0,0,0]                                 NaN   \n",
       "3                            [10,0,0,0]                                 NaN   \n",
       "4                            [0,3,0,12]                        [0, 8, 0, 8]   \n",
       "...                                 ...                                 ...   \n",
       "136242                      [0,0,117,0]                      [1, 0, 713, 0]   \n",
       "136243                      [0,266,0,0]                     [0, 1405, 0, 1]   \n",
       "136244                      [0,0,458,0]                     [0, 0, 1848, 0]   \n",
       "136245                              NaN                                 NaN   \n",
       "136246                              NaN                                 NaN   \n",
       "\n",
       "       bam-rc[A,C,G,T]_G27310.HCC1428.1 BaseCount[A,C,G,T]_G27249.AU565.1  \\\n",
       "0                            [4,0,12,0]                     [6, 0, 13, 0]   \n",
       "1                            [4,0,15,0]                     [6, 0, 11, 0]   \n",
       "2                             [1,0,0,0]                               NaN   \n",
       "3                             [7,0,0,0]                               NaN   \n",
       "4                             [0,8,0,8]                    [0, 28, 0, 15]   \n",
       "...                                 ...                               ...   \n",
       "136242                      [1,0,724,0]                   [3, 3, 2119, 1]   \n",
       "136243                     [0,1415,0,1]                   [0, 3837, 0, 1]   \n",
       "136244                     [0,0,1850,0]                   [3, 1, 6142, 1]   \n",
       "136245                        [0,0,0,0]                               NaN   \n",
       "136246                        [0,0,1,0]                               NaN   \n",
       "\n",
       "        ... BaseCount[A,C,G,T]_G26260.HCC1395.2  \\\n",
       "0       ...                                 NaN   \n",
       "1       ...                                 NaN   \n",
       "2       ...                                 NaN   \n",
       "3       ...                                 NaN   \n",
       "4       ...                       [0, 12, 0, 6]   \n",
       "...     ...                                 ...   \n",
       "136242  ...                       [0, 0, 24, 0]   \n",
       "136243  ...                       [0, 90, 0, 0]   \n",
       "136244  ...                      [0, 0, 165, 0]   \n",
       "136245  ...                                 NaN   \n",
       "136246  ...                                 NaN   \n",
       "\n",
       "       bam-rc[A,C,G,T]_G26260.HCC1395.2  \\\n",
       "0                             [5,0,1,0]   \n",
       "1                             [5,0,1,0]   \n",
       "2                             [0,0,0,0]   \n",
       "3                             [9,0,1,0]   \n",
       "4                            [0,13,0,6]   \n",
       "...                                 ...   \n",
       "136242                       [0,0,30,0]   \n",
       "136243                       [0,94,0,0]   \n",
       "136244                      [0,0,169,0]   \n",
       "136245                        [0,1,0,0]   \n",
       "136246                              NaN   \n",
       "\n",
       "       BaseCount[A,C,G,T]_G28072.MDA-MB-175-VII.1  \\\n",
       "0                                  [10, 0, 10, 0]   \n",
       "1                                   [8, 0, 19, 0]   \n",
       "2                                             NaN   \n",
       "3                                             NaN   \n",
       "4                                  [0, 27, 0, 19]   \n",
       "...                                           ...   \n",
       "136242                            [1, 1, 1304, 0]   \n",
       "136243                            [0, 2296, 0, 0]   \n",
       "136244                            [1, 0, 3501, 1]   \n",
       "136245                                        NaN   \n",
       "136246                                        NaN   \n",
       "\n",
       "       bam-rc[A,C,G,T]_G28072.MDA-MB-175-VII.1  \\\n",
       "0                                  [10,0,10,0]   \n",
       "1                                   [8,0,19,0]   \n",
       "2                                    [1,0,1,0]   \n",
       "3                                   [19,0,2,0]   \n",
       "4                                  [0,27,0,19]   \n",
       "...                                        ...   \n",
       "136242                            [1,1,1306,0]   \n",
       "136243                            [0,2297,0,0]   \n",
       "136244                            [1,0,3502,1]   \n",
       "136245                               [0,0,0,7]   \n",
       "136246                               [0,0,0,0]   \n",
       "\n",
       "       BaseCount[A,C,G,T]_G30624.UACC-812.1 bam-rc[A,C,G,T]_G30624.UACC-812.1  \\\n",
       "0                                       NaN                         [4,0,2,0]   \n",
       "1                                       NaN                         [4,0,1,0]   \n",
       "2                                       NaN                         [0,0,0,0]   \n",
       "3                                       NaN                         [4,0,0,0]   \n",
       "4                             [0, 7, 0, 10]                        [0,7,0,10]   \n",
       "...                                     ...                               ...   \n",
       "136242                       [0, 0, 370, 0]                       [0,0,373,0]   \n",
       "136243                       [0, 445, 0, 0]                       [0,451,0,0]   \n",
       "136244                       [0, 0, 565, 0]                       [0,0,566,0]   \n",
       "136245                                  NaN                               NaN   \n",
       "136246                                  NaN                               NaN   \n",
       "\n",
       "       BaseCount[A,C,G,T]_G20496.HCC1954.2 bam-rc[A,C,G,T]_G20496.HCC1954.2  \\\n",
       "0                                      NaN                        [0,0,5,0]   \n",
       "1                                      NaN                        [1,0,6,0]   \n",
       "2                                      NaN                        [2,0,0,0]   \n",
       "3                                      NaN                       [12,0,0,0]   \n",
       "4                            [0, 11, 0, 0]                       [0,11,0,0]   \n",
       "...                                    ...                              ...   \n",
       "136242                     [1, 0, 2900, 1]                     [1,0,2903,1]   \n",
       "136243                     [1, 4965, 0, 2]                     [1,4970,0,2]   \n",
       "136244                     [0, 0, 5800, 4]                     [0,0,5802,4]   \n",
       "136245                                 NaN                        [0,1,0,4]   \n",
       "136246                                 NaN                              NaN   \n",
       "\n",
       "       BaseCount[A,C,G,T]_G28069.MDA-MB-415.1  \\\n",
       "0                               [9, 0, 30, 0]   \n",
       "1                               [9, 0, 38, 0]   \n",
       "2                                [5, 0, 6, 0]   \n",
       "3                              [17, 0, 12, 0]   \n",
       "4                               [0, 12, 0, 1]   \n",
       "...                                       ...   \n",
       "136242                        [5, 2, 2367, 1]   \n",
       "136243                        [0, 4015, 0, 0]   \n",
       "136244                        [3, 0, 4944, 0]   \n",
       "136245                                    NaN   \n",
       "136246                                    NaN   \n",
       "\n",
       "       bam-rc[A,C,G,T]_G28069.MDA-MB-415.1  \n",
       "0                               [9,0,30,0]  \n",
       "1                               [9,0,38,0]  \n",
       "2                                [5,0,6,0]  \n",
       "3                              [17,0,12,0]  \n",
       "4                               [0,12,0,1]  \n",
       "...                                    ...  \n",
       "136242                        [5,2,2369,1]  \n",
       "136243                        [0,4019,0,0]  \n",
       "136244                        [3,0,4944,0]  \n",
       "136245                          [0,0,0,13]  \n",
       "136246                                 NaN  \n",
       "\n",
       "[136247 rows x 51 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt_df_final_minimal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La tabla anterior contiene las siguientes columnas:\n",
    "  * ***Region:*** Chromosome id.\n",
    "  * ***Position:*** Genomic Position.\n",
    "  * ***Reference:*** Base on reference genome.\n",
    "  * ***Basecount[A,C,G,T]:*** Base count from Reditools2 analysis.\n",
    "  * ***bam-rc[A,C,G,T]:*** Base count from bam-readcount analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dividir la tabla entre los 3 grupos de drogas\n",
    "\n",
    "YB, además requiere generar sub-sets que agrupan las líneas celulares según:\n",
    "  * La droga del tratamiento.\n",
    "  * Según si es respondedor o no respondedor a la droga. \n",
    "\n",
    "La tabla proporcionada se muestra a continuación:\n",
    "\n",
    "| Cell_line | PARPi | Anthracyclines | Alkylatingagents |\n",
    "|---:|---:|---:|---:|\n",
    "| AU565 | HS | HS | HS |\n",
    "| BT-474 | LS | LS | LS |\n",
    "| BT-483 | LS | LS | LS |\n",
    "| BT-549 | LS | HS | LS |\n",
    "| CAL-120 | LS | NaN | LS |\n",
    "| CAL-148 | HS | HS | HS |\n",
    "| CAL-51 | HS | HS | HS |\n",
    "| CAMA-1 | HS | HS | NaN |\n",
    "| EFM-192A | NaN | LS | LS |\n",
    "| HCC1419 | LS | LS | HS |\n",
    "| HCC1428 | LS | LS | LS |\n",
    "| HCC1500 | LS | NaN | HS |\n",
    "| HCC1937 | LS | LS | LS |\n",
    "| HCC1954 | HS | LS | LS |\n",
    "| HCC202 | NaN | LS | NaN |\n",
    "| HCC2218 | NaN | NaN | HS |\n",
    "| HCC38 | LS | NaN | LS |\n",
    "| HCC70 | HS | NaN | HS |\n",
    "| JIMT-1 | NaN | HS | NaN |\n",
    "| MCF7 | LS | HS | NaN |\n",
    "| MDA-MB-175-VII | LS | NaN | LS |\n",
    "| MDA-MB-231 | HS | HS | HS |\n",
    "| MDA-MB-436 | LS | HS | LS |\n",
    "| MDA-MB-468 | HS | HS | HS |\n",
    "| UACC-812 | HS | NaN | HS |\n",
    "| ZR-75-30 | LS | LS | LS |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_7225/1178716939.py:51: DtypeWarning: Columns (0,49) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mt = pd.read_csv('../outputs/05_mt-bamreadcount_merge/CCL24_master_table_just-counts_with_bamreadcount.tsv', sep=\"\\t\", header=0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------------------\n",
      "Saving file CCLE24_master_table_PARPi.tsv\n",
      "------------------------------------\n",
      "Saving file CCLE24_master_table_Anthracyclines.tsv\n",
      "------------------------------------\n",
      "Saving file CCLE24_master_table_Alkylatingagents.tsv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import glob\n",
    "\n",
    "# Function for re-filter the out table\n",
    "# Just filter by total coverage >= 10\n",
    "def filtering2(path_file, out_dir):\n",
    "    \n",
    "    # Create the out_file name\n",
    "    out_file = os.path.join(out_dir, os.path.basename(path_file).replace('.tsv', '_filtered.tsv'))\n",
    "    \n",
    "    count = 0\n",
    "    with open(path_file, 'r') as fp , open(out_file, 'w+') as fp_out:\n",
    "        for line in fp:\n",
    "            if count == 0:\n",
    "                fp_out.write(f\"{line}\")\n",
    "                count += 1\n",
    "                continue\n",
    "            count += 1\n",
    "            tmp = line.strip('\\n').split('\\t')\n",
    "            \n",
    "            line_filt = tmp[:3]\n",
    "            for col in tmp[3:]:\n",
    "                if col != 'NA':\n",
    "                    tmp2 = col.replace('[','').replace(']','').split(',')\n",
    "                    tmp2 = [int(x) for x in tmp2]\n",
    "                    if sum(tmp2) < 10:\n",
    "                        line_filt.append('NA')\n",
    "                    else:\n",
    "                        line_filt.append(col)\n",
    "                else:\n",
    "                    line_filt.append('NA')\n",
    "            line_filt2 = line_filt\n",
    "            line_filt2.append('\\n')\n",
    "            line_filt_out = '\\t'.join(line_filt2)\n",
    "            fp_out.write(f\"{line_filt_out}\")           \n",
    "    print('------------------------------------')\n",
    "\n",
    "\n",
    "\n",
    "# input metadata table and df reading\n",
    "table = \"../data/id_cell_lines.txt\"\n",
    "df = pd.read_csv(table, sep=\"\\t\", header=0)\n",
    "# Recode values\n",
    "df.replace('High-sensitivity','HS', inplace=True)\n",
    "df.replace('Low-sensitivity','LS', inplace=True)\n",
    "\n",
    "# Read the master table\n",
    "mt = pd.read_csv('../outputs/05_mt-bamreadcount_merge/CCL24_master_table_just-counts_with_bamreadcount.tsv', sep=\"\\t\", header=0)\n",
    "# Get column names\n",
    "mt_cols = mt.columns\n",
    "# Create a df with only the columns ['Region','Position','Reference'] and [bam-rc*].\n",
    "cols2keep = ['Region','Position','Reference']+[x for x in mt_cols if 'bam-rc' in x]\n",
    "#cols2keep2 = ['Region','Position','Reference']+[x for x in cols2keep if '_HS' in x]+[x for x in mt_cols if '_LS' in x]\n",
    "mt = mt[cols2keep]\n",
    "\n",
    "for drug in ['PARPi','Anthracyclines','Alkylatingagents']:\n",
    "    col_list = ['Region','Position','Reference']\n",
    "    col_new = ['Region','Position','Reference']\n",
    "    cell_list = list(df[~df[drug].isna()]['Cell_line'])\n",
    "    for cell in cell_list:\n",
    "        # get the group : HS or LS of the cell line on the current drug\n",
    "        group = df[df['Cell_line'] == cell][drug]\n",
    "        group = list(group)[0]\n",
    "\n",
    "        tmp_cols = [x for x in cols2keep if cell in x ]\n",
    "        new_cols = [x+'_'+group for x in cols2keep if cell in x]\n",
    "        \n",
    "        col_list = col_list+tmp_cols\n",
    "        col_new = col_new+new_cols\n",
    "\n",
    "    tmp_df = mt[col_list].copy()\n",
    "    tmp_df.columns = col_new\n",
    "\n",
    "    # Reorder columns\n",
    "    cols_order = ['Region','Position','Reference']+[x for x in col_new if '_HS' in x]+[x for x in col_new if '_LS' in x]\n",
    "    out_df = tmp_df[cols_order]\n",
    "\n",
    "    # Save tmp file for filtering\n",
    "    out_df.to_csv(os.path.join('../outputs/05_mt-bamreadcount_merge/', 'tmp.tsv'), sep=\"\\t\", header=True, index=False, na_rep='NA')\n",
    "    \n",
    "    # Re-filter the table\n",
    "    filtering2(os.path.join('../outputs/05_mt-bamreadcount_merge/', 'tmp.tsv'), '../outputs/05_mt-bamreadcount_merge/')\n",
    "    \n",
    "    \n",
    "    # Save table\n",
    "    outname = 'CCLE24_master_table_'+drug+'.tsv'\n",
    "    print('Saving file {}'.format(outname))\n",
    "    cmd = \"mv ../outputs/05_mt-bamreadcount_merge/tmp_filtered.tsv ../outputs/05_mt-bamreadcount_merge/{}\".format(outname)\n",
    "    os.system(cmd)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_7225/2652021181.py:10: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(table, sep=\"\\t\", header=0, na_values='NA')\n",
      "/tmp/ipykernel_7225/2652021181.py:10: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(table, sep=\"\\t\", header=0, na_values='NA')\n",
      "/tmp/ipykernel_7225/2652021181.py:10: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(table, sep=\"\\t\", header=0, na_values='NA')\n"
     ]
    }
   ],
   "source": [
    "# Add columns HS_count and LS_count.\n",
    "# Create list with path to tables\n",
    "tables = ['../outputs/05_mt-bamreadcount_merge/CCLE24_master_table_PARPi.tsv',\n",
    "          '../outputs/05_mt-bamreadcount_merge/CCLE24_master_table_Anthracyclines.tsv',\n",
    "          '../outputs/05_mt-bamreadcount_merge/CCLE24_master_table_Alkylatingagents.tsv']\n",
    "\n",
    "# For each table\n",
    "for table in tables:\n",
    "    # read the table as df\n",
    "    df = pd.read_csv(table, sep=\"\\t\", header=0, na_values='NA')\n",
    "    # get column names\n",
    "    cols = df.columns\n",
    "    # Get col names of HS and LS\n",
    "    cols_HS = [col for col in cols if \"_HS\" in col]\n",
    "    cols_LS = [col for col in cols if \"_LS\" in col]\n",
    "    df[\"HS_NAcount\"] = df[cols_HS].isnull().sum(axis=1)\n",
    "    df[\"HS_count\"] = df[cols_HS].notnull().sum(axis=1)    \n",
    "    df[\"LS_NAcount\"] = df[cols_LS].isnull().sum(axis=1)\n",
    "    df[\"LS_count\"] = df[cols_LS].notnull().sum(axis=1)\n",
    "\n",
    "    # Save the table\n",
    "    df.to_csv(table.replace('.tsv', '_groupCounts.tsv'), sep=\"\\t\", header=True, index=False, na_rep='NA')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FINAL TABLES\n",
    "\n",
    "Las tablas finales son :\n",
    "\n",
    "- `../outputs/05_mt-bamreadcount_merge/CCLE24_master_table_PARPi_groupCounts.tsv`\n",
    "- `../outputs/05_mt-bamreadcount_merge/CCLE24_master_table_Anthracyclines_groupCounts.tsv`\n",
    "- `../outputs/05_mt-bamreadcount_merge/CCLE24_master_table_Alkylatingagents_groupCounts.tsv`\n",
    "\n",
    "A esta tabla se le agregaron 4 columnas al final:\n",
    "\n",
    "* **HS_NAcount :**  \n",
    "            Counts of NA in the High Sensitivity Group\n",
    "\n",
    "* **HS_count :**  \n",
    "            Counts of VALID in the High Sensitivity Group \n",
    "\n",
    "* **LS_NAcount :**  \n",
    "            Counts of NA in the Low Sensitivity Group\n",
    "\n",
    "* **LS_count =**  \n",
    "            Counts of VALID in the Low Sensitivity Group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DO NOT USE: OLD CODE - TEST CODE - ATTEMPS "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counting reads with  Samtools\n",
    "Adicionalmente con ***Samtools*** se puede filtrar un poco los reads antes de contar, tal como se ve en el comando abajo:\n",
    "\n",
    "`samtools view -@ 10 -q 30 -F 3844 ${file} \"${region2}\"`\n",
    "\n",
    "> -q, --min-MQ INT           ...have mapping quality >= INT  \n",
    "> -F, --excl[ude]-flags FLAG ...have none of the FLAGs present\n",
    "\n",
    "Donde el FLAG 3844 significa:\n",
    "\n",
    "> read unmapped (0x4)  \n",
    "> not primary alignment (0x100)  \n",
    "> read fails platform/vendor quality checks (0x200)  \n",
    "> read is PCR or optical duplicate (0x400)  \n",
    "> supplementary alignment (0x800)  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "out_dir=\"../outputs/04_samtools_counts\"\n",
    "mkdir -p ${out_dir}\n",
    "\n",
    "# table with editions\n",
    "table=\"../outputs/02_master_table/CCLE24_master_table.tsv\"\n",
    "\n",
    "find ../data/bams -name \"*.bam\" | while read file; \n",
    "do \n",
    "    echo \"Analyzing ${file}\"\n",
    "    outname=$(basename ${file} | sed 's/Aligned.sortedByCoord.out.bam//g')\n",
    "    awk '{print $1\":\"$2\"-\"$3\"\\t\"$4\"\\t\"$5}' $enhancers  | while read region;\n",
    "    do\n",
    "        region2=$(echo \"$region\" | awk '{print $1}')\n",
    "        id=$(echo \"$region\" | awk '{print $2}' )\n",
    "        len=$(echo \"$region\" | awk '{print $3}')\n",
    "        count=$(samtools view -@ 10 -q 30 -F 3844 ${file} \"${region2}\" | wc -l)\n",
    "        echo -e \"${region2}\\t${id}\\t${len}\\t${count}\" >> tmp.samtools.count\n",
    "        \n",
    "    done        \n",
    "    mv tmp.samtools.count ${out_dir}/${outname}.counts\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
